### DISCUSSION TITLE
Module not Found Error in Angular

### DISCUSSION THREAD

**Discussion by aiira-co** [jrouwe/JoltPhysics.js#146] [Q&A] [answered]
Hello, i have decoupled my game engine physics which making it possible to switch engines: this was done in order to welcome jolt-physics.

I did the implementation with rapier to affirm that the decoupling works. switching to jolt, i kept receiving an error that Module could not be found. I have confirmed that the location and files are present in my node_modules, have reinstalled jolt-engine. 

`./node_modules/jolt-physics/dist/jolt-physics.wasm-compat.js:27:16-32 - Error: Module not found: Error: Can't resolve 'module' in 'C:\Code\InterverseEngine\node_modules\jolt-physics\dist'`

Wanted to know if there is a way around this. thank you

**Comment by jrouwe** [comment]
I'm sorry, I'm not familiar how things work in Angular.

I created an example using node.js [here](https://github.com/jrouwe/JoltPhysics.js/tree/main/helloworld) which seems to work just fine.

**Reply by aiira-co** [reply]
Thanks for your response. I will ask help from the angular community.

**Comment by PhoenixIllusion** [comment, marked as answer]
Potential Workaround:

I attempted a trivial angular build of the friction demo using the latest @angular/cli to check this out. While the regular `NPM run start` worked (I believe `serve` is now ViteJS powered), the `NPM run build` appears to error-out, related to the bundler parsing the jolt-physics code in a 'browser-only' context and hitting NodeJS-only code.

The jolt-physics emscripten-produced JS module code will be parsed for optimization, and it will see the existence of "if NodeJS's `process` exists, runtime-load the NodeJS's `module` package via import()". The bundler will not assume `process` will be undefined, so will not compile-out/ignore that segment of code. It therefore requires some way to resolve a JS package called 'module' during the bundling phase.

One way to make the Emscritpen module work in angular as-is is using the package.json's `browser` alternate package-map, and set `"module": false` in the package.json "browser" section to create a blank module:

(example: potentially putting this browser json at the bottom of package.json if one doesn't already exist)
```json
{
 ...
  "devDependencies": {
    ...
    "karma": "~6.4.0",
    "karma-chrome-launcher": "~3.2.0",
    "karma-coverage": "~2.2.0",
    "karma-jasmine": "~5.1.0",
    "karma-jasmine-html-reporter": "~2.1.0",
    "typescript": "~5.4.2"
  },
  "browser": {
    "module": false
  }
}
```

This will create am empty JS module called `module` during the bundling (a zero-byte file, likely named something `chunk-*.js` by default), and the rest of the code should work. It is kind of littering an extra file, but at runtime since the code still checks for `process` and sees it is undefined, so won't import the empty file.

Example loading the Friction demo, assumes typescript and uses a coarse conversion of `examples.js` to typescript:

```typescript
import { AfterViewInit, Component, ElementRef, ViewChild } from '@angular/core';
import { RouterOutlet } from '@angular/router';

import initJolt from 'jolt-physics/wasm-compat';
import { LAYER_MOVING, addToScene, bodyInterface, createFloor, initExample, setContainer } from './example';

@Component({
  selector: 'app-root',
  standalone: true,
  imports: [RouterOutlet],
  templateUrl: './app.component.html',
  styleUrl: './app.component.scss'
})
export class AppComponent implements AfterViewInit {
  title = 'jolt-angular';
  @ViewChild('container', {read: ElementRef})
  containerEl?: ElementRef<HTMLDivElement>;

  async ngAfterViewInit(): Promise<void> {
    setContainer(this.containerEl!.nativeElement);


    initJolt().then(async function (Jolt) {
      // Initialize this example
      initExample(Jolt, null);
    
      const floor = createFloor();
      floor.SetFriction(1.0);

      // Create objects with increasing friction
      for (let i = 0; i < 10; i++) {
        // Create physics body
        let shape = new Jolt.BoxShape(new Jolt.Vec3(0.5, 0.5, 0.5), 0.05);
        let creationSettings = new Jolt.BodyCreationSettings(shape, new Jolt.RVec3(-10.0 + 2.0 * i, 0.5, 0), Jolt.Quat.prototype.sIdentity(), Jolt.EMotionType_Dynamic, LAYER_MOVING);
        creationSettings.mFriction = 0.1 * i;
        let body = bodyInterface.CreateBody(creationSettings);

        // Apply an impulse to make the body move
        body.AddImpulse(new Jolt.Vec3(0, 0, 10000.0));

        addToScene(body, 0x00ffff);
      }
    })
  }
  
```

uses:
    "@types/three": "0.163.0",
    "three": "0.163.0",
    "jolt-physics": "0.22.0",
    "stats.js": "0.17.0"

`example.ts` here:
https://gist.github.com/PhoenixIllusion/791fc945e83a0ce3b3379b7c1596c69f

**Reply by PhoenixIllusion** [reply]
Using a disabled `module`, this should compile:
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/5e73f04a-dae3-4a19-8b6c-01d7fc7ef03c)

**Reply by aiira-co** [reply]
Thanks for your response, works for me.
